## Part 1

In [1]:
from nytimesarticle import articleAPI
api = articleAPI('tcrAZ92puwt0ApwLWpHkLjS9JIYb35SZ')
from bs4 import BeautifulSoup
from urllib.request import *
from nltk import word_tokenize
import nltk
from nltk.corpus import stopwords

In [2]:
articles = api.search( q = 'President Xi' )

In [3]:
mylink = []
print('\n--1--\n')
    
for element in articles:
    print(element)

print('\n--2--\n')
    
for element in articles['response']:
    print(element)

print('\n--3--\n')
    
for element in articles['response']['docs'][0]:
    print(element)

print('\n--4--\n')

print(articles['response']['docs'][0]['web_url'])
print(articles['response']['docs'][0]['snippet'])

print('\n--5--\n')

for element in articles['response']['docs']:
    mylink.append(element['web_url'])
    print(element['web_url'])


--1--

status
copyright
response

--2--

docs
meta

--3--

web_url
snippet
blog
source
multimedia
headline
keywords
pub_date
document_type
news_desk
byline
type_of_material
_id
word_count
score
uri

--4--

https://www.nytimes.com/2018/12/18/briefing/president-xi-kpop-huawei.html
Let us help you start your day.

--5--

https://www.nytimes.com/2018/12/18/briefing/president-xi-kpop-huawei.html
https://www.nytimes.com/2018/02/26/world/asia/xi-jinping-career-highlights.html
https://www.nytimes.com/2018/02/25/world/asia/xi-jinping-china.html
https://www.nytimes.com/2019/01/13/opinion/kim-jong-un-trump-china.html
https://www.nytimes.com/reuters/2019/01/23/technology/23reuters-china-markets-tech.html
https://www.nytimes.com/2018/12/09/opinion/trade-war-truce-trump-china.html
https://www.nytimes.com/2017/11/10/insider/the-last-time-president-xi-took-a-question-from-an-american-correspondent.html
https://www.nytimes.com/aponline/2019/01/21/world/asia/ap-as-china-cambodia.html
https://www.nytime

In [4]:
# save articles about president Xi to a dictionary, with title as key and content as value
mydic = {}
for link in mylink:
    xhtml = urlopen(link)
    xsource = xhtml.read()
    xsoup = BeautifulSoup(xsource, 'html.parser')
    content_raw = xsoup.find_all("div", {"class": "css-u5vfum StoryBodyCompanionColumn"})
    mycontent = []
    for p in content_raw:
        mycontent.append(p.get_text())
    mytitle = xsoup.head.title.get_text()
    mydic[mytitle] = mycontent

In [5]:
# check title
for title in mydic:
    print(title)

President Xi, K-Pop, Huawei: Your Wednesday Briefing - The New York Times
President Xi Jinping’s Rise in China, as Covered by The Times - The New York Times
President Xi’s Strongman Rule Raises New Fears of Hostility and Repression - The New York Times
Taiwan President Calls for International Support to Defend Democracy - The New York Times
Opinion | China Gives Kim Jong-un a Boost - The New York Times
U.S. Government Official Says Trade Talks With China Continuing Wednesday - The New York Times
Train Believed to Be Carrying North Korea's Kim Leaves Beijing - The New York Times
Opinion | Tete-a-tete after tit-for-tat - The New York Times
The Last Time President Xi Took a Question From an American Correspondent? 2014, and Yup, It Was Mine - The New York Times


In [12]:
mydic["Opinion | China Gives Kim Jong-un a Boost - The New York Times"]

['Leader Kim Jong-un of North Korea has met for a fourth time with China’s president, Xi Jinping. He’s still waiting for a second coveted summit with President Trump. With Beijing seemingly in his corner, and few words from the White House, Mr. Kim has mused about possibly finding “a new way” forward on denuclearization.Follow The New York Times Opinion section on Facebook, Twitter (@NYTopinion) and Instagram.']

In [6]:
# check content of one article
mydic['President Xi Jinping’s Rise in China, as Covered by The Times - The New York Times']

['HONG KONG — The announcement that the Chinese Communist Party intends to drop a two-term limit on the presidency is the latest sign that Xi Jinping is moving to strengthen the authority of the government and his position atop it.By upending a model of collective leadership that was put in place after the excesses of one-man rule and the power struggles that followed the deaths of Mao Zedong and Deng Xiaoping, President Xi has shown a determination to rewrite the rule book but also to reimpose older norms of a strong single leader ruling China. The prospect that he will rule for years to come has raised concerns about oppression at home and potential truculence abroad.Mr. Xi, the son of Communist Party elite, came to power with a long history of working his way up the ranks of officialdom, deftly sidelining rivals and rolling over dissent as he cultivated alliances and strengthened his control over the government, the economy and the military.Here is a guide to The New York Times’s co

In [7]:
# Use one article for this assignment
text = ""
for k,v in mydic.items():
    for p in v:
        text += p

In [8]:
# Reference https://stackoverflow.com/questions/14596884/remove-text-between-and-in-python
import re 
text = re.sub("([\[]).*?([\]])", "\g<1>\g<2>", text)
text = text.replace("■","").replace("•","").replace("[","").replace("]","") 
print(text)

(Want to get this briefing by email? Here’s the sign-up.)Good morning. President Xi Jinping’s unabashed defiance, K-pop’s unintended impact on air travel and Huawei’s aggressive corporate culture.  Xi Jinping doubles down. In a speech on the 40th anniversary of China’s economic liberalization efforts, President Xi, depicted above, delivered an unabashed defense of his economic policies.The Communist Party has been solely responsible for the country’s stunning growth, he argued, and its policies of control have been “totally correct.”He only obliquely referred to the elephant in the room: the trade war with the U.S. “No one is in a position to dictate to the Chinese people what should not be done,” he said.→ Takeaways: The speech doused hopes that Mr. Xi was ready to announce reforms that would make the economy more market-friendly and suggest potential compromises with the United States. Mr. Xi’s defiance drove Asian markets down._____ Trump Foundation to shut down. Adding to President

### Q1. In a few sentences, please describe the text (or corpus/corpora) you have chosen and any pre-processing steps (“cleaning”) you employed, along with a brief comment on why you chose it.


I used NYtimes API to search articles about President Xi and cleaned the searching results for this assignment. The contents of articles are quite clean. I only removed some bullet points from the text and words in "[]" as "[]" that are used to show the link address.

In [102]:
tokens = word_tokenize(text)
text = nltk.Text(tokens)

### Q2.	Provide the raw length of your text (in characters/letters).

In [108]:
len(text)

5493

### Q3. Provide the number of tokens in your text (use “set” to get the tokens).

In [109]:
len(set(text))

1805

### Q4.	Provide the lexical diversity score for your text, and comment on where yours falls relative to the texts we examined in Assignment 1 (and speculate why).

In [110]:
def lexical_diversity(text):   
    return len(set(text)) / len(text)

In [111]:
lexical_diversity(text)

0.3286000364099763

The score is close to text8 (Personals Corpus), 0.223. Both text8 and my chosen text include different people's article or corpus, which contributes to the lexcial richness.

### Q5.	Remove all punctuation from your text, and provide the raw length, number of tokens and lexical diversity score again.

In [119]:
# go a step further and eliminate numbers and punctuation from the vocabulary count by filtering out any non-alphabetic items
removedtext = [word.lower() for word in text if word.isalpha()]
print("The raw length of new text is", len(removedtext),'\n'
      "The number of tokens is", len(set(removedtext)),'\n'
      "The lexical diversity score is",lexical_diversity(removedtext) )

The raw length of new text is 4563 
The number of tokens is 1545 
The lexical diversity score is 0.33859303090072324


### Q6.	Remove all stopwords from your text (with the punctuation removed as in #5), and provide the raw length, number of tokens and lexical diversity score again.

In [123]:
stopwords = nltk.corpus.stopwords.words('english')
removedtext2 = [w for w in removedtext if w not in stopwords]
print("The raw length of new text is", len(removedtext2),'\n'
      "The number of tokens is", len(set(removedtext2)),'\n'
      "The lexical diversity score is",lexical_diversity(removedtext2) )

The raw length of new text is 2592 
The number of tokens is 1440 
The lexical diversity score is 0.5555555555555556


### Q7.	Comment on what the differences between the length, numbers of tokens and lexical diversity for your text with everything, without punctuation and without stopwords tell you about the text, and whether you found any of these results surprising or particularly interesting.

In [131]:
import nltk
from nltk.book import *

In [128]:
print("The raw length of text8 is", len(text8),'\n'
      "The number of tokens in text8 is", len(set(text8)),'\n'
      "The lexical diversity score of new text8 is",lexical_diversity(text8) )

The raw length of text8 is 4867 
The number of tokens in text8 is 1108 
The lexical diversity score of new text8 is 0.22765564002465585


In [127]:
removedtext8 = [word.lower() for word in text8 if word.isalpha()]
removedtext82 = [w for w in removedtext8 if w not in stopwords]
print("The raw length of new text8 is", len(removedtext82),'\n'
      "The number of tokens in new text8 is", len(set(removedtext82)),'\n'
      "The lexical diversity score of new text8 is",lexical_diversity(removedtext82) )

The raw length of new text8 is 2548 
The number of tokens in new text8 is 694 
The lexical diversity score of new text8 is 0.27237048665620095


Without punctuation and stopwords, the raw length and numbers of tokens decrease, but the lexical diversity score increases. The raw length reduces by half compared the raw text in Q6 (2592) with the anwser to Q2 (5493). The number of tokens also reduces from 1805 to 1440. The lexical diversity increases from 0.3286000364099763 to 0.5555555555555556. I expect to see the increase of lexical diversity score when I removed punctuation and stopwords, but I am suprised to see such a high score (0.55). So I do the same to text8 as text 8 and my chosen text have similar lexical diversity score (Q4). The changes of lexical diversity score for text8 is not that big, only from 0.23 to 0.27. The reason of the difference between my text and text 8 could be articles on nytimes are more professional, complex and diverse.

### Q8.	Provide all words in your text with lengths greater than 12 letters.

In [130]:
V = set(removedtext2)
long_words = [w for w in V if len(w) > 12]
sorted(long_words)

['administration',
 'affectionately',
 'anticorruption',
 'circumstances',
 'congressional',
 'consolidation',
 'constitutional',
 'controversial',
 'conversations',
 'correspondents',
 'cybersecurity',
 'demonstration',
 'determination',
 'globalization',
 'infrastructure',
 'international',
 'investigators',
 'liberalization',
 'manufacturing',
 'misrepresented',
 'reunification',
 'revolutionary',
 'semiautomatic',
 'semiautonomous',
 'traditionally',
 'unconventional',
 'understanding',
 'unpredictable']

### Q9.	Provide the output of the “unusual_words_many_no_endings_s_ed_ing” function (see page 2)  for your text.  Comment on what these words tell you about your text.

In [132]:
from nltk.corpus import words
from nltk.corpus import wordnet 

def unusual_words(text):
    text_vocab = set(w.lower() for w in text if w.isalpha())
    english_vocab = set(w.lower() for w in nltk.corpus.words.words())
    unusual = text_vocab - english_vocab
    return sorted(unusual)

def unusual_words_many(text):
    text_vocab = set(w.lower() for w in text if w.isalpha())
    manywords = nltk.corpus.words.words() + list(nltk.corpus.wordnet.words())
    english_vocab = set(w.lower() for w in manywords)
    unusual = text_vocab - english_vocab
    return sorted(unusual)

def unusual_words_many_no_endings_s_ed_ing(text):
    ut1 = unusual_words(text)
    print(str(len(ut1))+" words not in words.words\n")
    ut2 = unusual_words_many(text)
    print(str(len(ut2))+" words not in words.words + wordnet.words\n")
    ut3 = sorted(set(w.lower() for w in ut2 if not w.endswith("s") and not w.endswith("ed") and not w.endswith("ing")))
    print(str(len(ut3))+" words not in words.words + wordnet.words + no -s, -ed and -ing :\n")
    print(ut3)
    # -en plurals, -er and -est still not uncommon


In [133]:
unusual_words_many_no_endings_s_ed_ing(removedtext2)

414 words not in words.words

295 words not in words.words + wordnet.words

62 words not in words.words + wordnet.words + no -s, -ed and -ing :

['anticorruption', 'aung', 'barack', 'became', 'began', 'broader', 'children', 'cna', 'cybersecurity', 'deng', 'durex', 'facebook', 'finkelstein', 'flynn', 'francese', 'fujian', 'hannah', 'huawei', 'huckabee', 'instagram', 'jacqueline', 'jett', 'jiang', 'jintao', 'kong', 'kyi', 'largest', 'liu', 'macfie', 'mossberg', 'mào', 'màoyì', 'nytopinion', 'obama', 'ren', 'renmin', 'roantree', 'ryan', 'róngyì', 'schultz', 'selfie', 'smarter', 'softer', 'storychina', 'stronger', 'tiananmen', 'tsai', 'wray', 'xiaobo', 'yearslong', 'yimou', 'yinhong', 'yongkang', 'yì', 'zedong', 'zemin', 'zhejiang', 'zhongxun', 'zhàn', '容易', '易', '贸']


Some words are Chinese Pinyin ('mào'), Chinese lastname ('liu') and Chinese characters ('容易'). Some are brand name ('huawei') and place names ('zhejiang'). Some are different morphologies of words ('softer'). Others are uncommon words ('cybersecurity') or created by formatting problems ('storychina').

### Q10.	Provide the most common 50 words in your text (the 50 “most frequent”) after punctuation and stopwords have been removed.  Comment on whether any of them are surprising to you and what they indicate about the text.

In [134]:
fdist = FreqDist(removedtext2)
fdist.most_common(50)

[('xi', 67),
 ('china', 44),
 ('said', 35),
 ('chinese', 34),
 ('president', 31),
 ('trump', 24),
 ('would', 15),
 ('new', 15),
 ('power', 14),
 ('party', 13),
 ('trade', 13),
 ('obama', 13),
 ('jinping', 12),
 ('one', 11),
 ('also', 11),
 ('beijing', 11),
 ('political', 10),
 ('two', 10),
 ('communist', 9),
 ('first', 9),
 ('people', 8),
 ('united', 8),
 ('states', 8),
 ('security', 8),
 ('house', 8),
 ('times', 8),
 ('leader', 8),
 ('questions', 8),
 ('country', 7),
 ('media', 7),
 ('officials', 7),
 ('year', 7),
 ('white', 7),
 ('might', 7),
 ('government', 7),
 ('world', 7),
 ('could', 7),
 ('rule', 7),
 ('taiwan', 7),
 ('tsai', 7),
 ('press', 7),
 ('get', 6),
 ('economic', 6),
 ('war', 6),
 ('like', 6),
 ('leaders', 6),
 ('great', 6),
 ('support', 6),
 ('question', 6),
 ('make', 5)]

There are only one unexpected word, "White". I am curious that how articles about Chinese president discuss White group or the color white?

### Q11.	Compare the most common 50 words in your text to the most common 50 words in one of the 9 built-in texts we used in Assignment 1, with punctuation and stopwords removed.  Comment on any similarities or differences you note, and what these similarities and differences indicate about the texts themselves.


In [136]:
fdist8 = FreqDist(removedtext82)
fdist8.most_common(50)

[('lady', 88),
 ('seeks', 72),
 ('male', 42),
 ('looking', 34),
 ('ship', 33),
 ('slim', 33),
 ('fun', 31),
 ('attractive', 29),
 ('relationship', 29),
 ('build', 27),
 ('good', 26),
 ('seeking', 25),
 ('non', 25),
 ('smoker', 23),
 ('n', 23),
 ('guy', 22),
 ('honest', 22),
 ('movies', 22),
 ('age', 21),
 ('married', 21),
 ('like', 20),
 ('friendship', 20),
 ('single', 19),
 ('rship', 19),
 ('dining', 19),
 ('fit', 19),
 ('r', 18),
 ('caring', 17),
 ('f', 17),
 ('tall', 16),
 ('interests', 16),
 ('enjoy', 16),
 ('likes', 16),
 ('times', 15),
 ('open', 15),
 ('genuine', 15),
 ('female', 15),
 ('meet', 15),
 ('life', 15),
 ('fship', 14),
 ('man', 14),
 ('gsoh', 14),
 ('music', 14),
 ('would', 14),
 ('social', 14),
 ('employed', 13),
 ('ft', 13),
 ('drinker', 13),
 ('poss', 13),
 ('medium', 12)]

I still used text 8 for this question. The most common words for my chosen text are mainly about the heatedly discussed topics in political science and China, whereas the common words for text 8 are more about life and hobies as text 8 is a personal corpus.